pkg/workloads/containerd/watcher.go

```diff
@@ -52,82 +51,72 @@ const (
 	eventQueueBufferSize = 100
 )
 
-// containerEvents holds per-container queues for events
-type containerEvents struct {
-	sync.Mutex
-	events map[string]chan dTypesEvents.Message
-}
-
-func (ce *containerEvents) enqueueByContainerID(e dTypesEvents.Message) {
-	ce.Lock()
-	defer ce.Unlock()
-
-	if _, found := ce.events[e.Actor.ID]; !found {
-		q := make(chan dTypesEvents.Message, eventQueueBufferSize)
-		ce.events[e.Actor.ID] = q
-		go processContainerEvents(e.Actor.ID, q)
-	}
-	ce.events[e.Actor.ID] <- e
-}
-
-func (ce *containerEvents) reapEmpty() {
-	ce.Lock()
-	defer ce.Unlock()
-
-	for id, q := range ce.events {
-		if len(q) == 0 {
-			close(q)
-			delete(ce.events, id)
-		}
-	}
-}
-
 func shortContainerID(id string) string {
 	return id[:10]
 }
 
 // EnableEventListener watches for docker events. Performs the plumbing for the
 // containers started or dead.
 func EnableEventListener() error {
-	eventQueue := containerEvents{events: make(map[string]chan dTypesEvents.Message)}
 	since := time.Now()
-	syncWithRuntime()
-
-	eo := dTypes.EventsOptions{Since: strconv.FormatInt(since.Unix(), 10)}
-	r, err := dockerClient.Events(ctx.Background(), eo)
-	if err != nil {
+	ws := newWatcherState()
+	syncWithRuntime(ws, since)
+	if err := startEventListener(ws, since); err != nil {
 		return err
 	}
 
-	go listenForDockerEvents(&eventQueue, r)
-
 	// start a go routine which periodically synchronizes containers
 	// managed by the local container runtime and checks if any of them
 	// need to be managed by Cilium. This is a fall back mechanism in case
 	// an event notification has been lost.
 	go func() {
+		currentWatcherState := ws // inherit this from EnableEventListener's scope
 		for {
 			time.Sleep(syncRateDocker)
-			syncWithRuntime()
+			since := time.Now()
+			newWS := newWatcherState()
+			if outOfSync := syncWithRuntime(newWS, since); outOfSync {
+				log.Debugf("Out of sync, restarting containerd watcher with fresh state")
+				stopEventListener(currentWatcherState)
+				currentWatcherState = newWS
+				startEventListener(newWS, since)
+				return
```

Comment:
I think this will make the the out of sync feature to only work once :-)

Reply:
yup, it will! thanks for catching it. (I've removed it)

Reply:
I reshuffled the order of the loop, since in the case we resynced, we would then run the `reapEmpty` on the new state. Generally harmless but it might be confusing, so it now reaps first.